### ISSUE TITLE
Improve update detection for .NET (Core) applications deployed with ClickOnce

### ISSUE THREAD

**Issue by NikolaMilosavljevic** [dotnet/deployment-tools#27]
.NET (Core) applications deployed with ClickOnce cannot use `ApplicationDeployment` class for on-demand update check.

Besides checking for new updates, this check can be used to ensure that the deployment is still available on the server and wasn't deleted. If application is doing silent updates, without this check, user won't be aware that application is not getting any updates.

Launcher could do the update check if application is configured to install new updates automatically. We need to find a way to communicate the result of this check to .NET (Core) application. We could potentially also include the result of any recent automatic update.

**Comment by Rmurray0809** [comment]
Would there be any reason ClickOnce couldn't just drop a JSON file with information into the application directory when it is started?  It seems like this would allow an easy way for the application find out information about the install (ie put the version thats on the server, the install URL etc).  It could have extra information in it that would allow the developer to take action based on the information (exceptions?).  This would also make it fairly extensible so if someone wanted to add information in the future they could and generally applications will just continue to work.

**Comment by NikolaMilosavljevic** [comment]
@Rmurray0809 yes, a similar approach is considered - we'll have more details on the design soon.

**Comment by TimKras** [comment]
Any updates about this?

**Comment by NikolaMilosavljevic** [comment]
@TimKras thank you for the interest in this feature. We don't have an estimate, yet, when it would be available as a preview.

cc @dleeapho

**Comment by derskythe** [comment]
For those people who want to use ClickOnce on .NET 5 and want silently update, I wrote a small wrapper that implements some of the ApplicationDeployment properties:

- CurrentVersion
- ServerVersion
- UpdateAvailable
- Update
- IsNetworkDeployment
- DataDir

Hope this helps someone: https://github.com/derskythe/WpfSettings

**Comment by greatoceansoftware** [comment]
I definitely need the functionality of ApplicationDeployment. I can't seeing declaring victory of bringing ClickOnce to .NET 5/6 without it.

This is mainly because the limitation of only checking at application startup is annoying for most users. It's funny that they don't mind a splash screen with loading progress, but they mind this. Mainly because they want update checking AT THEIR CONVENIENCE. And I agree in the case of most of my applications.

Thanks for any consideration. Without it, it's back to commercial installers.

**Comment by jonrmorgan99** [comment]
Totally agree - without in-app updating ClickOnce is neutered. MSIX is not a real world alternative at present. Under NetFX ClickOnce proved it's reliability which is presumably why it made the cut to .Net5/6. 

So please re-open this issue and let's have a progress update. Thanks.

**Comment by luronumen** [comment]
Unable to port WinForms project from **.NET Framework 4.8** to **.NET 5.0** due to the following issue:
_CS0234 - The type or namespace name 'Deployment' does not exist in the namespace 'System' (are you missing an assembly reference?)_

**Comment by jonrmorgan99** [comment]
**@derskythe** 

> For those people who want to use ClickOnce on .NET 5 and want silently update, I wrote a small wrapper that implements some of the ApplicationDeployment properties


Thanks for posting this workaround for the missing ApplicationDeployment methods in Net5. You mention that the workaround features Silent Updates. However on my setup OpenURL(setuppath) in your Update code boots my browser and offers to download the application manifest manually. Any idea how I can get the app to update silently ?

Thanks.

**Comment by Mrxx99** [comment]
This is on the 6.0 milestone, so will this be fixed until next week?

**Comment by DDD25583** [comment]
I just tried to migrate a net4.8 app to .NET 6 when I came across this issue.
Being able to get the CurrentVersion is crucial in are error and bug-processing procedures.
As the production environment contains a lot of computers, updating (close and restart) an app is done once in the 24 hours.
So it is important that we know on which version of the app, an error is thrown.

**Comment by simmotech** [comment]
Based on ideas/code from @derskythe, @alansingfield, @e-master, I created a repro at https://github.com/simmotech/Net6ClickOnce which passes through all the info from args/ApplicationDeployment/ActivationArguments to the .NET 6 ClickOnce app. It also allows in-app detection of any later version and updating to that version on close.

Might be worth a look to use until an official fix is made. Grateful for any suggestions/further testing.

**Comment by jonrmorgan99** [comment]
Great to see ClickOnce in Net6, but still no progress on restoring the in-app updating functions in Application.Deployment.

Is there any update on this issue ? Net 7 maybe ??

**Comment by raffaeler** [comment]
@NikolaMilosavljevic Do you have any news for this?
We are still stuck since https://github.com/dotnet/deployment-tools/issues/9#issuecomment-726207037
This really prevents the migration of a large number of applications.

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic Do you have any news for this? We are still stuck since [#9 (comment)](https://github.com/dotnet/deployment-tools/issues/9#issuecomment-726207037) This really prevents the migration of a large number of applications.

We are definitely tracking improvements in ClickOnce experience for .NET (Core) apps, for upcoming release. You can expect to see some designs published in the next several weeks. Improvements might be brought up in stages, to allow for a better design and experience, as well as more testing with early previews.

**Comment by raffaeler** [comment]
@NikolaMilosavljevic Thanks for the prompt answer.

I noticed that the new publishing wizard creates far more files than in the past. The applications file are copied in both the publish folder and the sub-folder that is versioned.
Why does this differ from the past? I expected to see basically the same set of files, even because we host these files in a web application were there are both old and new ClickOnce apps.

**Comment by John-Hart** [comment]
The new ClickOnce for .NET publish wizard is essentially built on top of the standard .NET Folder Publish, after it runs and copies the files to the output folder, ClickOnce publish creates the manifest, setup.exe, launcher.exe and publish.html and then it copies the application files into the Application Files folder.  So that does mean there is an extra copy of the application files left in the output folder.

**Comment by jonrmorgan99** [comment]
Thanks for the explanation John.

 

It is really tedious to have to weed out the extra copy of the application files each time you do a Net6 CO build. 

Is there any possibility of getting ClickOnce in Net6 to delete these unneeded files ? 

 

As has been pointed out the Netfx CO build format was nice and clean – just the Application files folder, manifest, setup and publish.

**Comment by John-Hart** [comment]
@jonrmorgan99  we do have a feature suggestion to remove the extra files, feel free to vote on https://developercommunity.visualstudio.com/t/Clickonce-for-NET-Core-issues/1500772 if you would like to see that feature. 

In the meantime you might be able to try using the powershell [remove-item](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/remove-item?view=powershell-7.2) command like this:

    remove-item * -exclude *.application,launcher.exe,setup.exe,publish.html,'Application Files'

**Comment by berets76** [comment]
> 

Is it so difficult to give a definitive answer about System.Deployment in NET >= 6 ?
Please, simply tell us if we will get back IsNetworkDeployed and other properties and functions or not.

We don't need exotic solutions or dangerous shell-based workarounds, please just say yes or no, so we can evaluate alternatives with our customers.

Thanks

**Comment by NikolaMilosavljevic** [comment]
> > 
> 
> Is it so difficult to give a definitive answer about System.Deployment in NET >= 6 ? Please, simply tell us if we will get back IsNetworkDeployed and other properties and functions or not.
> 
> We don't need exotic solutions or dangerous shell-based workarounds, please just say yes or no, so we can evaluate alternatives with our customers.
> 
> Thanks

We plan to add support for IsNetworkDeployed and other properties that were previously exposed with https://docs.microsoft.com/en-us/dotnet/api/system.deployment.application.applicationdeployment?view=netframework-4.8

Properties will be available to .NET app as process environment variables. A simple class could be implemented, by the app developer, to abstract those variables and make their usage easier and very similar to old ApplicationDeployment class. We will provide the sample usage as part of the PR for this work, and in documentation.

The work is happening during 7.0 dotnet-mage development.

**Comment by jonrmorgan99** [comment]
@NikolaMilosavljevic Thanks for the update. 

What's needed for users hoping to port NetFx ClickOnce apps to Net6/7 is a complete rebuild of the ApplicationDeployment class for Net6/7, not just a few properties.

We need to be able to replicate the code example in

[ https://docs.microsoft.com/en-us/dotnet/api/system.deployment.application.applicationdeployment?view=netframework-4.8](url) 

which is the de-facto standard for implementing in-app updating with ClickOnce. 

This means exposing events such as CheckforUpdateCompleted, CheckForUpdateProgressChanged and UpdateCompleted as well as methods such as CheckForUpdate/CheckForUpdateAsync  and Update/UpdateAsync.

Please explain how process environment variables could be used to replace this essential functionality. As @berets76 commented, shell-based workarounds raise security risks.

ApplicationDeployment is an elegant, simple to use, robust solution to in-app updating with ClickOnce.

Why fix it if it ain't broke ?

Thanks

**Comment by NikolaMilosavljevic** [comment]
@jonrmorgan99, in-app update is not something that can simply be enabled, for ClickOnce-deployed .NET apps. There is no ClickOnce runtime in .NET (Core), with which the app can interact and initiate in-app update.

Launcher (built as .NET FX app) was added as a necessary entry point that 'hooks-up' to ClickOnce runtime. That model provides automatic updates, update check before launch, and, with upcoming addition, exposes various app properties, like URL parameters.

I understand that there is convenience, and perhaps a need for some types of applications, in checking for updates and performing update step while application is running. The app would still need to be restarted in the end. New properties will provide easy to consume insight into application Uri. This will enable a deterministic app restart using some of dfshim exports, i.e. LaunchApplication or ShOpenVerbApplication. App restart would let it download the pending update before launch.

The only missing piece is how to determine if there is a pending update available at deployment location. While there are no ready to use APIs at the moment, your application could do this check in a more direct way by determining the version that is available on the server. Certainly not ideal, but an alternative that would work.

**Comment by UFuz** [comment]
> Might be worth a look to use until an official fix is made. Grateful for any suggestions/further testing.

I just tried your code. So I assume, I need to copy "launcher.exe" over the one generated by VST? Doing this, when installing from the deploy location, I run into "Reference in the manifest does not match the identity of the downloaded assembly Launcher.exe." error. I am using the built in signing process. Can you or someone reading this, provide me steps how to re-sign this in an automated way? Unfortunately I am not experienced in doing this manually.

**Comment by alansingfield** [comment]
@UFuz - you've probably copied the Launcher.exe AFTER the manifest files have been generated.
The manifest contains a SHA256 hash of the executable, so if you amend the Launcher.exe the hash won't match.

There is also some other weirdness in that the behaviour of "dotnet mage -New Application" changes significantly depending on if there is a file called Launcher.exe alongside your main exe or not.
If not, it uses the .NET 4.8 protocol where every file is SHA256 hashed.
If Launcher.exe exists, I think it only hashes Launcher.exe itself.
Don't quote me on that, it's a while since I did this...

Have a look at these:
https://github.com/dotnet/deployment-tools/issues/149
https://github.com/alansingfield/ClickOncePowershell

**Comment by orinsdev** [comment]
Hello! Will this be available on .NET 7? Thanks!

**Comment by UFuz** [comment]
> If you just need an "Update Now" button in your app, you can shell out to this command and then exit.
> 
> ```batchfile
> rundll32 dfshim.dll,ShOpenVerbApplication http://yourupdateurl/yourapp.application
> ```

@alansingfield : With original .Net4.8 update mechanism, I used to run 
ApplicationDeployment updateCheck = ApplicationDeployment.CurrentDeployment;
updateCheck.Update();
show a success messagebox;
Application.Restart();

with the "dfshim.dll" approach I have 3 issues:
1) it returns immediately, not waiting for the update to complete
2) application.restart will relaunch the old version, instead of the new
3) restarting by desktop icon does not maintain the application settings, so app starts with defaults 

Is there a way to improve this, especially point 3?

**Comment by heku** [comment]
except the issue discussed above, I have another question, anybody know why a .net 6 WPF application with "online" mode only, cannot update itself when a new version published？（the application is not signed, but I didn't see any doc says update detect needs sign）

**Comment by NikolaMilosavljevic** [comment]
> except the issue discussed above, I have another question, anybody know why a .net 6 WPF application with "online" mode only, cannot update itself when a new version published？（the application is not signed, but I didn't see any doc says update detect needs sign）

@heku was the application built using dotnet-mage or ClickOnce publishing in Visual Studio?

cc @sujitnayak @John-Hart

**Comment by sujitnayak** [comment]
@NikolaMilosavljevic  The VS ClickOnce behavior in .NET 6.0 apps is consistent with .NET 4.X. Update is not enabled if app is Online only.

**Comment by heku** [comment]
@sujitnayak visual studio, it is totally different as my expectation, I thought Online mode means always “latest”, if so, what the purpose of Online mode？

**Comment by John-Hart** [comment]
@heku this might help: https://learn.microsoft.com/en-us/visualstudio/deployment/how-to-specify-the-clickonce-offline-or-online-install-mode?view=vs-2022

**Comment by heku** [comment]
@John-Hart Thx, I saw the doc before, it only says the difference of online/offline is online needs access to publish location, never mentions any difference about update behavior

**Comment by sujitnayak** [comment]
Behavior of auto-update not being available in online-only mode is not new to .NET 6.0 apps. .NET 4.X apps have had the same behavior as you can see in the screenshot attached:

![image](https://user-images.githubusercontent.com/7729842/195956378-78425d5c-7eaa-48a2-8b6e-cf0ff057042a.png)

Auto-update is for apps that are installed locally on the machine (visible in Add Remove Programs and Start menu) which happens only when offline mode is also available for the app.

**Comment by heku** [comment]
@sujitnayak ok, the "updates" button is disabled for Online mode, I thought it means Online mode app is always up to date, that what I really misunderstand. since it has same behavior as .NET Framework, I have no question any more, thank you.

**Comment by NikolaMilosavljevic** [comment]
ClickOnce support for .NET ships separately from main .NET (i.e. .NET 7). It ships as a global tool. We will be releasing version 7.0.0 of dotnet-mage in the next few days.

This feature is available in our preview release: https://www.nuget.org/packages/Microsoft.DotNet.Mage/7.0.0-preview.2.22317.2. For all previews you'd need to specify explicit version number, to install it, following the steps for the tool installation from the doc page: https://github.com/dotnet/deployment-tools/tree/main/Documentation/dotnet-mage

Please note that this feature is already available in Visual Studio 2022 build 17.3 or later.

Here's the relevant PR with more details about this feature and some examples: https://github.com/dotnet/deployment-tools/pull/208

**Comment by tejs-code** [comment]
@NikolaMilosavljevic Is there an alternate to UpdateAsync,  a way to install the update in the backgroun? I understand the app has to be restarted anyway, but it provides a way to download and process update in the background without paying a cost on open and showing a dialog to the user whether to install the update or not.

**Comment by TimKras** [comment]
@NikolaMilosavljevic 
How should this feature used in an application.
Do we need to read the `Environment.GetEnvironmentVariable("ClickOnce_IsNetworkDeployed")` manually, or will there be a wrapper? Or should we use this class in our application: https://github.com/dotnet/deployment-tools/blob/dfb0a98c68da39fc9cd1d82e200a19badf4e7112/Documentation/dotnet-mage/ApplicationDeployment.cs

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic How should this feature used in an application. Do we need to read the `Environment.GetEnvironmentVariable("ClickOnce_IsNetworkDeployed")` manually, or will there be a wrapper? Or should we use this class in our application: https://github.com/dotnet/deployment-tools/blob/dfb0a98c68da39fc9cd1d82e200a19badf4e7112/Documentation/dotnet-mage/ApplicationDeployment.cs

Yes, at the moment, those would be the only 2 solutions. You could use that sample wrapper class or create a new own. Reading environment variables would also work, if you only need to use a few.

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic Is there an alternate to UpdateAsync, a way to install the update in the backgroun? I understand the app has to be restarted anyway, but it provides a way to download and process update in the background without paying a cost on open and showing a dialog to the user whether to install the update or not.

There is no alternative today, for silent update, which was provided by UpdateAsync in the past (for .NET FX apps). That experience requires running update in the app context, which isn't possible today as System.Deployment library is not available in .NET.

We understand that this is an important scenario for some applications and will be looking into ways to close this gap in experience.

**Comment by luronumen** [comment]
Hi @NikolaMilosavljevic 

I noticed that the [ApplicationDeployment.cs](https://github.com/dotnet/deployment-tools/blob/dfb0a98c68da39fc9cd1d82e200a19badf4e7112/Documentation/dotnet-mage/ApplicationDeployment.cs) class still missing the 2 methods that are available on ClickOnce for .NET 4.8 Framework:
- CheckForDetailedUpdate()
- Update()

Is there any plan to implement them or any alternative methods to them on .NET 7.0?

Thanks in advanced,
Luciano

**Comment by NikolaMilosavljevic** [comment]
> Hi @NikolaMilosavljevic
> 
> I noticed that the [ApplicationDeployment.cs](https://github.com/dotnet/deployment-tools/blob/dfb0a98c68da39fc9cd1d82e200a19badf4e7112/Documentation/dotnet-mage/ApplicationDeployment.cs) class still missing the 2 methods that are available on ClickOnce for .NET 4.8 Framework:
> 
> * CheckForDetailedUpdate()
> * Update()
> 
> Is there any plan to implement them or any alternative methods to them on .NET 7.0?
> 
> Thanks in advanced, Luciano

That is correct. The sample class adds support for obtaining properties and does not have any methods related to updating the app. We understand that this is an important scenario for some applications and will be looking into ways to close this gap in experience.

**Comment by luronumen** [comment]
> That is correct. The sample class adds support for obtaining properties and does not have any methods related to updating the app. We understand that this is an important scenario for some applications and will be looking into ways to close this gap in experience.

Thank you very much for your prompt reply @NikolaMilosavljevic !
Are these gaps planned to be resolved in .NET 7.0 or will they only be available in .NET 8.0?

Thanks again!

**Comment by NikolaMilosavljevic** [comment]
Those features are not planned for 7.0, which GA'd last week. dotnet-mage, which will GA in a day or two, always support building deployments for older .NET versions, unless that is not feasible. For instance, new features that we add in dotnet-mage version 8 would support .NET 8.0, and all previous .NET releases.

**Comment by jonrmorgan99** [comment]
Hi @NikolaMilosavljevic

 Just following up @luronumen  and your response:

>     I noticed that the [ApplicationDeployment.cs](https://github.com/dotnet/deployment-tools/blob/dfb0a98c68da39fc9cd1d82e200a19badf4e7112/Documentation/dotnet-mage/ApplicationDeployment.cs) class still missing the 2 methods that are available on ClickOnce for .NET 4.8 Framework:
> 
>         CheckForDetailedUpdate()
>         Update()
> 
>     Is there any plan to implement them or any alternative methods to them on .NET 7.0?
> 
>     Thanks in advanced, Luciano
> 
> That is correct. The sample class adds support for obtaining properties and does not have any methods related to updating the app. We understand that this is an important scenario for some applications and will be looking into ways to close this gap in experience.

Has there been any progress in restoring methods for updating ClickOnce apps ? Is there a timeline to include them in .NET 8.0 ?
Is there anything the user community can do to help ?

Thanks for continuing to develop ClickOnce !

**Comment by mike7ang1rdz** [comment]
> ClickOnce support for .NET ships separately from main .NET (i.e. .NET 7). It ships as a global tool. We will be releasing version 7.0.0 of dotnet-mage in the next few days.
> 
> This feature is available in our preview release: https://www.nuget.org/packages/Microsoft.DotNet.Mage/7.0.0-preview.2.22317.2. For all previews you'd need to specify explicit version number, to install it, following the steps for the tool installation from the doc page: https://github.com/dotnet/deployment-tools/tree/main/Documentation/dotnet-mage
> 
> Please note that this feature is already available in Visual Studio 2022 build 17.3 or later.
> 
> Here's the relevant PR with more details about this feature and some examples: #208

are you guys not releasing a non preview version to a LTS NET version?.  We are currently releasing to NET 6 and due to short-term support not releasing to NET 7.

you should also target to NET 6 because it is LTS.

**Comment by mike7ang1rdz** [comment]
> > Hi @NikolaMilosavljevic
> > I noticed that the [ApplicationDeployment.cs](https://github.com/dotnet/deployment-tools/blob/dfb0a98c68da39fc9cd1d82e200a19badf4e7112/Documentation/dotnet-mage/ApplicationDeployment.cs) class still missing the 2 methods that are available on ClickOnce for .NET 4.8 Framework:
> > 
> > * CheckForDetailedUpdate()
> > * Update()
> > 
> > Is there any plan to implement them or any alternative methods to them on .NET 7.0?
> > Thanks in advanced, Luciano
> 
> That is correct. The sample class adds support for obtaining properties and does not have any methods related to updating the app. We understand that this is an important scenario for some applications and will be looking into ways to close this gap in experience.


 CheckForDetailedUpdate
this method is stopping me any workarounds available like calling service directly?

**Comment by statler** [comment]
Hi @NikolaMilosavljevic

Any update on the support for methods from the ApplicationDeployment class? Or somewhere there is a proposal we can track for this? I understand the technical challenges with the .Net/Core changes compared to framework, but surely there is a way...

**Comment by JDA88** [comment]
This issue (and the lack of workarounds) is preventing us to migrate to .NET7

**Comment by berets76** [comment]
Could we just have an idea of if/when this functionality will be restored?
This is a very important feature in some applications, we have been stalling for months, but now we need a certain time target.

Please just say yes or no, so we can possibly take other ways.
If it won't even be present in version 8, we will necessarily have to find an alternative.

Thanks

**Comment by NikolaMilosavljevic** [comment]
There are no immediate plans to add official support for in-app update functionality. There are some, perhaps less-than-ideal workarounds, provided in the comments of this issue. It is worth exploring and fine-tuning those options today.

.NET applications can access Application deployment properties using methods described in this PR: https://github.com/dotnet/deployment-tools/pull/208

Besides here at GitHub, I would also suggest voicing these and similar requests through Developer Community site: https://developercommunity.visualstudio.com/home

**Comment by fatih252** [comment]
Hello @NikolaMilosavljevic , any update or possible planning yet to support the extra methods ?

**Comment by DarrenNavitas** [comment]
Has there been any further update to this?

Is Clickonce technolody ultimately deprecated ?  If so, is there a newer recommended deployment / update mechanism, or is it a case of making your own?

**Comment by statler** [comment]
Darren - I have been moving between the available supported options from MS for about 10 years, and sad to say there really is no good MS supported option for .NET and never really has been. Clickonce is great when it works, but it remains flakey on update detection on some machines (as it ever was) and not an option in some network configurations. MSI still works, with all its issues and clunky update options it is usually bullet-proof if inconvenient. My conclusion is you have to either make peace with clickonce as it is (it _does_ work in most cases), or commit to taking it into your own hands or that of a 3rd party.

**Comment by DarrenNavitas** [comment]
Thanks @statler - to be honest we've been using ClickOnce for years - and we're perfectly happy with it.  The standard update methodology works fine for 90% of our use cases - even with new .net 8 apps where they are manually launched by the user. It's just we have a new .net 8 app that sits in the system tray doing it's thing - so on this occasion we'd want to use the ApplicationDeployment class and methods to have the app periodically check for any updates and install them - which is where the lack of support for Click Once is causing a slight pain !!

**Comment by statler** [comment]
Pretty much the same for us. It works really well for basic cases. Our main issue is when its automatic versioning detection doesn't work and the app just doesn't update itself, even though there are new versions available. Unfortunately this happens for some machines and not others and on some networks and not others, and it doesn't seem to be recoverable outside a reinstall. A manual 'update to latest' would sort this for us.

**Comment by nuzolx** [comment]
@statler When you say you have applications that won't update, is this associated with a crash when attempting to update? Which leads to the need for a clean installation. 
Or is it simply that the new version is not detected?

**Comment by statler** [comment]
Just not reliably detected.

**Comment by DarrenNavitas** [comment]
Not sure if its exactly the same, but we've experienced that - and found it depends on the shortcut being used to launch the app.  Sometimes if the users launch the tool - and then when running pin the icon to the taskbar  - that shortcust can be launching the app but bypassing the clickonce check.  We had to make sure they adding a shortcut from the .appref-ms file - which ensured the ClickOnce check is undertaken before launching.
